feat(ModularForms): Monotonicity of F/G (Proposition 8.12)

cameronfreer opened the pull request:
## Summary

This PR proves **Blueprint Proposition 8.12**: the function `FmodGReal(t) = F(it)/G(it)` is strictly decreasing on `(0, ∞)`.

### Main results (`FG.lean`)

* `G_vanishing_order` : `G / q^(3/2) → 20480` as `im(z) → ∞`
* `D_G_div_G_tendsto` : `(D G)/G → 3/2` as `im(z) → ∞`
* `L₁₀_imag_axis_real` : `L₁,₀(it)` is real for all `t > 0`
* `L₁₀_div_FG_tendsto` : `L₁,₀(it) / (F(it)G(it)) → 1/2` as `t → ∞`
* `L₁₀_eventually_pos_imag_axis` : `L₁,₀(it) > 0` for large `t`
* `L₁₀_pos` : `L₁,₀(it) > 0` for all `t > 0` (via `antiSerreDerPos`)
* `FmodGReal_differentiableOn` : `FmodGReal` is differentiable on `(0, ∞)`
* `deriv_FmodGReal` : derivative formula `(-2π) * L₁₀(it) / G(it)²`
* `deriv_FmodGReal_neg` : `deriv FmodGReal t < 0` for all `t > 0`
* `FmodG_strictAntiOn` : `FmodGReal` is strictly decreasing on `(0, ∞)`

### Supporting infrastructure

**Derivative.lean:**
* `MDifferentiable_div` : Division of MDifferentiable functions (when denominator nonvanishing)
* `D_tendsto_zero_of_isBoundedAtImInfty` : D-derivative of bounded holomorphic function → 0 at infinity (Cauchy estimate)

**JacobiTheta.lean:**
* `differentiableAt_jacobiTheta₂_half` : Differentiability of `t ↦ jacobiTheta₂(t/2, t)`
* `Θ₂_MDifferentiable` : MDifferentiability of `Θ₂`

**FG.lean helpers:**
* `logderiv_mul_eq` : `D(f·h)/(f·h) = Df/f + Dh/h` (pointwise, where defined)
* `div_re_of_im_eq_zero` : `(a/b).re = a.re / b.re` when `b` is real-valued
* `D_cexp_div` : `D(exp(c·z))/exp(c·z) = c/(2πi)`
* `H₂_eventually_ne_zero` : Eventual nonvanishing of `H₂`
* `D_B_tendsto_zero` : Cauchy-estimate vanishing for the B polynomial
* `qexp_tendsto_of_poly_bound` : Generic q-expansion limit from polynomial coefficient bounds

### Sorry status

All new theorems are sorry-free. Pre-existing sorries remain in:
* `F_eq_FReal`, `G_eq_GReal`, `FmodG_eq_FmodGReal` (sorry'd connectors)
* `FReal_Differentiable`, `GReal_Differentiable` (future work)
* `MLDE_F`, `MLDE_G` (modular linear differential equations)
* `FmodG_rightLimitAt_zero`, `FG_inequality_1`, `FG_inequality_2` (future work)

## Dependencies

All merged: #259, #269, #270

seewoo5 commented on SpherePacking/ModularForms/Derivative.lean:
```diff
@@ -320,6 +475,46 @@ example : D (E₄.toFun * E₄.toFun) = 2 * 3⁻¹ * E₄.toFun * (E₂ * E₄.t
+- Since D = (2πi)⁻¹ · d/dz, we have F' = 2πi · D F
+- So d/dt F(it) = 2πi · D F(it) · I = -2π · D F(it)
+-/
+theorem deriv_resToImagAxis_eq (F : ℍ → ℂ) (hF : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) F)
```
I think you can separate this theorem as a single PR (in the same file). I guess it may also shorten the proof of `antiDerPos` suggested in #213 (which is in review).

cameronfreer replied:
Thanks, I've separated this as #249 (but not yet refactored #239 to build on top of that).

cameronfreer commented:
@seewoo5 I think I'm done with my initial pass on this. As noted in the PR description, it might be easiest to wait for #193, #251, and #252 to merge first. Do you have suggestions on what would be easiest to break off of this first for review after those? Thanks.

I'm not sure if #248 will need to merge first, though that might make things simpler. Also #213 is needed for the transitive dependencies of `Q_strictAntiOn` to be sorry-free, but I don't think this PR needs to wait on it.

One possibility is to go in order of files:
  | PR  | Content                                                                           | ~Lines | Dependencies             |
  |-----|-----------------------------------------------------------------------------------|--------|--------------------------|
  | A1  | Derivative.lean extensions (D_qexp, Ramanujan scaffolding)                        | ~300   | None (or perhaps #248) |
  | A2  | FG.lean extensions (Δ_fun_eq_Δ, L₁₀ def) + Monotonicity_ImagAxis.lean             | ~400   | A1                       |
  | A3  | Monotonicity_L10_SerreAndF.lean (F-side analysis)                                 | ~950   | A2                       |
  | A4  | Monotonicity_L10_GAndEventuallyPos.lean + Monotonicity.lean (G-side + final glue) | ~600   | A3                       |

Alternatively, we could divide it up by proof structure:
  | PR  | Content                                                                       | Dependencies |
  |-----|-------------------------------------------------------------------------------|--------------|
  | B1  | Imaginary axis positivity/realness (Monotonicity_ImagAxis.lean)               | FG defs      |
  | B2  | Serre derivative positivity (serre_D_L₁₀_pos_imag_axis from L10_SerreAndF)    | B1           |
  | B3  | Eventual positivity (L₁₀_eventually_pos_imag_axis from L10_GAndEventuallyPos) | B1           |
  | B4  | Final glue (Monotonicity.lean: Q, deriv_Q, Q_strictAntiOn)                    | B2, B3       |
  
I'm also happy to follow some other approach if you prefer.

cameronfreer commented:
awaiting-review

cameronfreer commented:
The build is expected to fail until #269 and #270 are merged. (Lean can't resolve the imports and so there will be "unknown module" errors.)

seewoo5 commented on SpherePacking/ModularForms/FG.lean:
```diff
@@ -871,8 +882,716 @@ theorem D_F_div_F_tendsto :
+  exact h_lim.congr fun z => by ring
+
+/-- `D(2H₂² + 5H₂H₄ + 5H₄²) → 0` as `im(z) → ∞`. -/
+private theorem D_B_tendsto_zero :
```
I think we can just keep this here - it is shorter than I thought.

cameronfreer commented on SpherePacking/ModularForms/FG.lean:
```diff
@@ -871,12 +882,556 @@ theorem D_F_div_F_tendsto :
+-/
+
+/-- `L₁,₀(it) > 0` for all `t > 0`. -/
+theorem L₁₀_pos_imag_axis : ResToImagAxis.Pos L₁₀ := L₁₀_pos
```
Oops, thanks: a5790e069e7512573cc45ec05bf6fd6887cde5b6

cameronfreer commented on SpherePacking/ModularForms/FG.lean:
```diff
@@ -871,8 +882,716 @@ theorem D_F_div_F_tendsto :
 end AsymptoticAnalysis
 
+/- $\mathcal{L}_{1, 0}$ is eventually positive on the imaginary axis. -/
+lemma L₁₀_eventuallyPos : ResToImagAxis.EventuallyPos L₁₀ := L₁₀_eventually_pos_imag_axis
```
See f0e0f2fc11ec4ff74761edcc1816e6a4dacba76c

seewoo5 commented on SpherePacking/ModularForms/FG.lean:
```diff
@@ -871,13 +882,539 @@ theorem D_F_div_F_tendsto :
+/-- `D(2H₂² + 5H₂H₄ + 5H₄²) → 0` as `im(z) → ∞`, by the Cauchy estimate. -/
+private theorem D_B_tendsto_zero :
+    Filter.Tendsto (fun z : ℍ =>
+      D (fun w => 2 * H₂ w ^ 2 + 5 * H₂ w * H₄ w + 5 * H₄ w ^ 2) z)
```
Can you replace this and similar occurrences with point-free versions? I keep seeing this in your PRs and wonder if these are necessary or can be changed.

cameronfreer commented:
@seewoo5 I think I made all the changes above, but please let me know if there's more.

I also extracted `MDifferentiable_div` to simplify `hh_md` -- it doesn't save tons of lines but makes things more modular and might be useful elsewhere? See [83e74d8](https://github.com/thefundamentaltheor3m/Sphere-Packing-Lean/pull/239/commits/83e74d8a28a6c22037eb6d420c13b291a4fa1cc6).

cameronfreer commented:
also golfed a bit more in 93bff09:


  1. logderiv_mul_eq (line 138): Merged rw + simp only into single simp only (-1 line)
  2. div_re_of_im_eq_zero (line 142-144): Used Complex.div_ofReal_re instead of manually converting both a and b to ofReal. Also removed unnecessary ha
  hypothesis — only hb is needed (-2 lines, -1 param)
  3. qexp_arg_imag_axis_pnat (line 152-157): Dropped explicit type annotation for intermediate have, using exp_imag_axis_arg t ht n directly (-2 lines)
  4. D_cexp_div (line 876-884): Collapsed two nested have blocks into single term-mode .trans chain, combined rw + field_simp on one line (-3 lines)
